File: algorithm/枚举/小于n的最大素数/sieve_of_Eratosthenes.py
# 埃拉托斯特尼筛法
import math

n = 100000

a = {i: True for i in range(1, n + 1)}

for x in range(2, int(math.sqrt(n)) + 1):
    for y in range(2, int(n / x) + 1):
        if x * y <= n:
            a[y * x] = False

last = 2

for key, value in a.items():
    if value:
        last = key

print(last)  # 0.28s user 0.01s system 98% cpu 0.292 total

File: algorithm/枚举/小于n的最大素数/sieve_of_Eratosthenes.rs
//0.04s user 0.00s system 96% cpu 0.041 total
fn main() {
    let n:i32 = 1000000;

    let mut a = [true; 1000000];

    for x in 2..((n as f32).sqrt() as i32 + 1) {
        for y in 2..(n/x+1 as i32) {
            if x*y-1 <= n {
                a[(x*y-1) as usize] = false;
            }
        }
    }

    let mut last = 2;
    for (key, value) in a.iter().enumerate() {
        if *value {
            last = key+1;
        }
    }
    println!("{}", last);
}
